leo/commands/setup.rs

```diff
@@ -64,75 +61,75 @@ impl Command for Setup {
-                    proving_key.vk.write(&mut verification_key)?;
-                    let _ = verification_key_file.write_to(&path, &verification_key)?;
-                    tracing::info!("Complete");
-
-                    (proving_key, prepared_verifying_key)
-                } else {
-                    tracing::info!("Detected saved setup");
-
-                    // Read the proving key file from the output directory
-                    tracing::info!("Loading proving key...");
-
-                    if self.skip_key_check {
-                        tracing::info!("Skipping curve check");
-                    }
-                    let proving_key_bytes = ProvingKeyFile::new(&package_name).read_from(&path)?;
-                    let proving_key =
-                        Parameters::<Bls12_377>::read(proving_key_bytes.as_slice(), !self.skip_key_check)?;
-                    tracing::info!("Complete");
-
-                    // Read the verification key file from the output directory
-                    tracing::info!("Loading verification key...");
-                    let verifying_key_bytes = VerificationKeyFile::new(&package_name).read_from(&path)?;
-                    let verifying_key = VerifyingKey::<Bls12_377>::read(verifying_key_bytes.as_slice())?;
-
-                    // Derive the prepared verifying key file from the verifying key
-                    let prepared_verifying_key = PreparedVerifyingKey::<Bls12_377>::from(verifying_key);
-                    tracing::info!("Complete");
-
-                    (proving_key, prepared_verifying_key)
-                };
-
-                Ok((program, proving_key, prepared_verifying_key))
-            }
-            None => {
-                let mut main_file_path = path;
-                main_file_path.push(SOURCE_DIRECTORY_NAME);
-                main_file_path.push(MAIN_FILENAME);
+        if input.is_none() {
+            return Err(anyhow!("Unable to build, check that main file exists"));
+        }
 
-                Err(anyhow!("Unable to build, check that main file exists"))
+        let (program, checksum_differs) = input.unwrap();
+
+        // Check if a proving key and verification key already exists
+        let keys_exist = ProvingKeyFile::new(&package_name).exists_at(&path)
+            && VerificationKeyFile::new(&package_name).exists_at(&path);
+
+        // If keys do not exist or the checksum differs, run the program setup
+        // If keys do not exist or the checksum differs, run the program setup
+        let (proving_key, prepared_verifying_key) = if !keys_exist || checksum_differs {
+            tracing::info!("Starting...");
+
+            // Run the program setup operation
+            let (proving_key, prepared_verifying_key) = {
+                let rng = &mut thread_rng();
+                let setup_res = Groth16::<Bls12_377, Compiler<Fr, _>, Vec<Fr>>::setup(&program, rng);
+                match setup_res {
```

Comment:
@howardwu 

Also this change is important - I match `setup_res` and catch error. If `?` is used or error directly returned - stack overflow happens.